src/PostgREST/MediaType.hs

```diff
@@ -114,48 +117,58 @@ toMimePlanFormat PlanText = "text"
+        "application/vnd.pgrst.plan":rest        -> getPlan PlanText rest
+        "application/vnd.pgrst.plan+text":rest   -> getPlan PlanText rest
+        "application/vnd.pgrst.plan+json":rest   -> getPlan PlanJSON rest
+        "application/vnd.pgrst.object+json":rest -> checkSingularNullStrip rest
+        "application/vnd.pgrst.object":rest      -> checkSingularNullStrip rest
+        "application/vnd.pgrst.array+json":rest  -> checkArrayNullStrip rest
+        "application/vnd.pgrst.array":rest       -> checkArrayNullStrip rest
+        "*/*":_                                  -> MTAny
+        other:_                                  -> MTOther other
+        _                                        -> MTAny
+      where
+        checkArrayNullStrip ["nulls=stripped"] = MTVndArrayJSONStrip
+        checkArrayNullStrip  _                 = MTApplicationJSON
 
-    checkSingularNullStrip ["nulls=stripped"] = MTVndSingularJSON True
-    checkSingularNullStrip  _                 = MTVndSingularJSON False
+        checkSingularNullStrip ["nulls=stripped"] = MTVndSingularJSON True
+        checkSingularNullStrip  _                 = MTVndSingularJSON False
 
-    getPlan fmt rest =
-      let
-        opts         = BS.split (BS.c2w '|') $ fromMaybe mempty (BS.stripPrefix "options=" =<< find (BS.isPrefixOf "options=") rest)
-        inOpts str   = str `elem` opts
-        dropAround p = BS.dropWhile p . BS.dropWhileEnd p
-        mtFor        = fromMaybe MTApplicationJSON $ do
-          foundFor    <- find (BS.isPrefixOf "for=") rest
-          strippedFor <- BS.stripPrefix "for=" foundFor
-          pure . decodeMediaType $ dropAround (== BS.c2w '"') strippedFor
-      in
-      MTVndPlan mtFor fmt $
-        [PlanAnalyze  | inOpts "analyze" ] ++
-        [PlanVerbose  | inOpts "verbose" ] ++
-        [PlanSettings | inOpts "settings"] ++
-        [PlanBuffers  | inOpts "buffers" ] ++
-        [PlanWAL      | inOpts "wal"     ]
+        getPlan fmt rest =
+          let
+            opts         = T.splitOn "|" $ fromMaybe mempty (T.stripPrefix "options=" =<< find (T.isPrefixOf "options=") rest)
+            inOpts str   = str `elem` opts
+            dropAround p = T.dropWhile p . T.dropWhileEnd p
+            mtFor        = fromMaybe MTApplicationJSON $ do
+              foundFor    <- find (T.isPrefixOf "for=") rest
+              strippedFor <- T.stripPrefix "for=" foundFor
+              pure . decodeMediaType' $ dropAround (== '"') strippedFor
+          in
+          MTVndPlan mtFor fmt $
+            [PlanAnalyze  | inOpts "analyze" ] ++
+            [PlanVerbose  | inOpts "verbose" ] ++
+            [PlanSettings | inOpts "settings"] ++
+            [PlanBuffers  | inOpts "buffers" ] ++
+            [PlanWAL      | inOpts "wal"     ]
+    normalizeMediaType :: ByteString -> Text
+    normalizeMediaType t = T.toLower $ decodeLatin1 t -- Latin1 is sufficient for HTTP headers
```

Comment:
> Decode a [ByteString](https://hackage.haskell.org/package/bytestring-0.11.5.2/docs/Data-ByteString.html#t:ByteString) containing Latin-1 (aka ISO-8859-1) encoded text.

https://hackage.haskell.org/package/text-2.1.1/docs/Data-Text-Encoding.html#v:decodeLatin1

> ISO-8859-1 was the default encoding of the values of certain descriptive HTTP headers

https://en.wikipedia.org/wiki/ISO/IEC_8859-1

> The default charset of ISO-8859-1 for text media types has been
   removed; the default is now whatever the media type definition says.

https://datatracker.ietf.org/doc/html/rfc7231#appendix-B

Seems ISO-8859-1 is not 100% correct but looks sufficient for now. How about some comments clarifying this? Explicitly mentioning `ISO-8859-1` would be good.

Reply:
@wolfgangwalther https://www.rfc-editor.org/rfc/rfc5987 defines the way to encode text of arbitrary charset in HTTP headers, but that should be handled somewhere down the web stack, maybe by Warp itself even.

Dropped the `FIXME` comment altogether BTW.